internal/controller/sink.go

```diff
@@ -0,0 +1,302 @@
+package controller
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	gwapiv1 "sigs.k8s.io/gateway-api/apis/v1"
+	"sigs.k8s.io/yaml"
+
+	aigv1a1 "github.com/envoyproxy/ai-gateway/api/v1alpha1"
+	"github.com/envoyproxy/ai-gateway/extprocconfig"
+)
+
+const selectedBackendHeaderKey = "x-envoy-ai-gateway-selected-backend"
+
+// configSinkEvent is the interface for the events that the configSink can handle.
+// It can be either an LLMBackend, an LLMRoute, or a deletion event.
+type configSinkEvent any
+
+// configSinkEventLLMBackendDeleted is an event to notify the configSink that an LLMBackend has been deleted.
+type configSinkEventLLMBackendDeleted struct{ namespace, name string }
+
+// configSinkEventLLMRouteDeleted is an event to notify the configSink that an LLMRoute has been deleted.
+type configSinkEventLLMRouteDeleted struct{ namespace, name string }
+
+// configSink centralizes the LLMRoute and LLMBackend objects handling
+// which requires to be done in a single goroutine since we need to
+// consolidate the information from both objects to generate the ExtProcConfig
+// and HTTPRoute objects.
+type configSink struct {
+	client client.Client
+	kube   kubernetes.Interface
+	logger logr.Logger
+
+	eventChan                   chan configSinkEvent
+	httpRoutes                  map[string]*gwapiv1.HTTPRoute
```

Comment:
Seems like we are maintaining the local cache here, we can consider using controller runtime client which supports cached client to primarily retrieve data from in-memory cache that's automatically updated by Kubernetes informers., write operations still directly interact with APIServer.
https://github.com/kubernetes-sigs/controller-runtime/tree/main/pkg/client

Reply:
yeah intentionally manually maintaining the cache here - we can iterate on this here or in another PR later. I will take a look if i get a time tomorrow but this can be a TODO i guess